io-sim/src/Control/Monad/IOSim.hs

```diff
@@ -266,3 +332,138 @@ ppEvents events =
+    cache = unsafePerformIO $ newIORef $
+              -- we use opts here just to be sure the reference cannot be
+              -- lifted out of exploreSimTrace
+              if explorationScheduleBound opts>=0
+                then Set.empty
+                else error "exploreSimTrace: negative schedule bound"
+    cached m = unsafePerformIO $ atomicModifyIORef' cache $ \set ->
+      (Set.insert m set, Set.member m set)
+    cacheSize () = unsafePerformIO $ Set.size <$> readIORef cache
+
+replaySimTrace :: forall a test. (Testable test)
+               => ExplorationOptions
+               -> (forall s. IOSim s a)
+               -> ScheduleControl
+               -> (Maybe (SimTrace a) -> SimTrace a -> test)
+               -> Property
+replaySimTrace opts mainAction control k =
+  let (_,trace) = detachTraceRaces $
+                            controlSimTrace (explorationStepTimelimit opts) control mainAction
+      in property (k Nothing trace)
+
+raceReversals :: ScheduleControl -> Int
+raceReversals ControlDefault      = 0
+raceReversals (ControlAwait mods) = length mods
+raceReversals ControlFollow{}     = error "Impossible: raceReversals ControlFollow{}"
+
+-- compareTraces is given (maybe) a passing trace and a failing trace,
+-- and identifies the point at which they diverge, where it inserts a
+-- "sleep" event for the thread that is delayed in the failing case,
+-- and a "wake" event before its next action. It also returns the
+-- identity and time of the sleeping thread. Since we expect the trace
+-- to be consumed lazily (and perhaps only partially), and since the
+-- sleeping thread is not of interest unless the trace is consumed
+-- this far, then we collect its identity only if it is reached using
+-- unsafePerformIO.
+
+compareTraces :: Maybe (SimTrace a1)
+              -> SimTrace a2
+              -> (Maybe ((Time, ThreadId, Maybe ThreadLabel),
+                         Set.Set (ThreadId, Maybe ThreadLabel)),
+                  SimTrace a2)
+compareTraces Nothing trace = (Nothing, trace)
+compareTraces (Just passing) trace = unsafePerformIO $ do
+  sleeper <- newIORef Nothing
+  return (unsafePerformIO $ readIORef sleeper,
+          go sleeper passing trace)
+  where go sleeper (SimTrace tpass tidpass _ _ pass')
+           (SimTrace tfail tidfail tlfail evfail fail')
+          | (tpass,tidpass) == (tfail,tidfail) =
+              SimTrace tfail tidfail tlfail evfail $
+                go sleeper pass' fail'
+        go sleeper (SimTrace tpass tidpass tlpass _ _) fail =
+          unsafePerformIO $ do
+            writeIORef sleeper $ Just ((tpass, tidpass, tlpass),Set.empty)
+            return $ SimTrace tpass tidpass tlpass EventThreadSleep $
+                       wakeup sleeper tidpass fail
+        go _ _ fail = fail
+        wakeup sleeper tidpass (SimTrace tfail tidfail tlfail evfail fail')
+          | tidpass == tidfail =
+              SimTrace tfail tidfail tlfail EventThreadWake fail'
```

Comment:
We are ignoring `evfail`, is it ok?  Shouldn't we just add `EventThreadWake`, e.g. return
```haskell
     SimTrace tfail itdfail tlfail EventThreadWake
   $ SimTrace tfail tidfail tlfail evfail
   $ fail'
```

Reply:
Yes, you're absolutely right.